File: aat-android/src/main/java/ch/bailu/aat/map/mapsforge/MapsForgePreview.java
package ch.bailu.aat.map.mapsforge;


import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Color;

import org.mapsforge.core.graphics.Canvas;
import org.mapsforge.core.model.BoundingBox;
import org.mapsforge.core.model.Dimension;
import org.mapsforge.core.model.MapPosition;
import org.mapsforge.core.model.Point;
import org.mapsforge.map.android.graphics.AndroidGraphicFactory;
import org.mapsforge.map.layer.Layer;
import org.mapsforge.map.layer.TilePosition;
import org.mapsforge.map.util.LayerUtil;
import org.mapsforge.map.util.MapPositionUtil;
import org.mapsforge.map.view.FrameBuffer;
import org.mapsforge.map.view.FrameBufferHA2;

import java.io.OutputStream;
import java.util.List;

import ch.bailu.aat.map.MapDensity;
import ch.bailu.aat.map.tile.TileProvider;
import ch.bailu.aat.map.tile.source.CacheOnlySource;
import ch.bailu.aat_lib.service.cache.DownloadSource;
import ch.bailu.aat_lib.map.tile.source.MapsForgeSource;
import ch.bailu.aat_lib.map.tile.source.Source;
import ch.bailu.aat.preferences.Storage;
import ch.bailu.aat.preferences.map.SolidMapTileStack;
import ch.bailu.aat_lib.preferences.map.SolidRenderTheme;
import ch.bailu.aat.util.OldAppBroadcaster;
import ch.bailu.aat.util.graphic.SyncTileBitmap;
import ch.bailu.aat_lib.app.AppContext;
import ch.bailu.aat_lib.dispatcher.AppBroadcaster;
import ch.bailu.aat_lib.gpx.GpxInformation;
import ch.bailu.aat_lib.gpx.InfoID;
import ch.bailu.aat_lib.logger.AppLog;
import ch.bailu.aat_lib.map.layer.gpx.GpxDynLayer;
import ch.bailu.aat_lib.service.directory.MapPreviewInterface;
import ch.bailu.foc.Foc;

public class MapsForgePreview extends MapsForgeViewBase implements MapPreviewInterface {
    private static final int BITMAP_SIZE=128;
    private static final Dimension DIM = new Dimension(BITMAP_SIZE, BITMAP_SIZE);
    private static final Source MAPNIK = new CacheOnlySource(DownloadSource.MAPNIK);

    private final Foc          imageFile;
    private final TileProvider provider;


    private final MapPosition mapPosition;
    private final BoundingBox bounding;
    private final Point tlPoint;

    public MapsForgePreview(Context context, AppContext appContext, GpxInformation info, Foc out) throws IllegalArgumentException {
        super(context, appContext.getServices(), MapsForgePreview.class.getSimpleName(), new MapDensity());



        layout(0, 0, BITMAP_SIZE, BITMAP_SIZE);
        getModel().mapViewDimension.setDimension(DIM);

        imageFile = out;
        provider = new TileProvider(appContext, getSource(new SolidRenderTheme(appContext.getMapDirectory(), appContext)));

        MapsForgeTileLayer tileLayer = new MapsForgeTileLayer(appContext.getServices(), provider);
        add(tileLayer, tileLayer);

        GpxDynLayer gpxLayer = new GpxDynLayer(new Storage(getContext()), getMContext(), appContext.getServices());
        add(gpxLayer);

        attachLayers();

        gpxLayer.onContentUpdated(InfoID.FILEVIEW, info);
        frameBounding(info.getGpxList().getDelta().getBoundingBox());

        mapPosition = getModel().mapViewPosition.getMapPosition();
        int tileSize = getModel().displayModel.getTileSize();
        bounding = MapPositionUtil.getBoundingBox(mapPosition, DIM, tileSize);
        tlPoint = MapPositionUtil.getTopLeftPoint(mapPosition, DIM, tileSize);


        preLoadTiles();

    }

    private void preLoadTiles() {
        List<TilePosition> tilePositions = LayerUtil.getTilePositions(bounding,
                mapPosition.zoomLevel, tlPoint,
                getModel().displayModel.getTileSize());

        provider.onAttached();
        provider.preload(tilePositions);
    }


    private static Source getSource(SolidRenderTheme stheme) {
        SolidMapTileStack tiles = new SolidMapTileStack(stheme);

        boolean[] enabled = tiles.getEnabledArray();

        if (enabled[0]) {

            String theme = stheme.getValueAsString();


            MapsForgeSource mfs = new MapsForgeSource(theme);
            return new CacheOnlySource(mfs);
        }

        return MAPNIK;
    }


    /**
     *
     * Begin of "prevent MapView from drawing" hack
     * FIXME:
     *   This hack prevents the map view from calling the layers draw() function.
     *   The correct implementation would be to port the preview generator away from the MapView and
     *   just use the MapViews model.
     */

    @Override
    public FrameBuffer getFrameBuffer() {
        // TODO: change this to HA3 when updating to MapsForge 0.15
        return new FrameBufferHA2(getModel().frameBufferModel, getModel().displayModel, AndroidGraphicFactory.INSTANCE) {
            @Override
            public org.mapsforge.core.graphics.Bitmap getDrawingBitmap() {
                return null;
            }
        };
    }

    @Override
    public void repaint() {}

    @Override
    public void requestRedraw() {}
    /**
     * End of "prevent MapView from drawing" hack
     */


    private SyncTileBitmap generateBitmap() {
        final SyncTileBitmap bitmap = new SyncTileBitmap();

        bitmap.set(BITMAP_SIZE, false);
        if (bitmap.getAndroidBitmap() != null) {
            final android.graphics.Canvas c = bitmap.getAndroidCanvas();
            final Canvas canvas = AndroidGraphicFactory.createGraphicContext(c);

            bitmap.getAndroidBitmap().eraseColor(Color.BLACK);

            for (Layer layer : getLayerManager().getLayers()) {
                layer.draw(bounding, mapPosition.zoomLevel, canvas, tlPoint);
            }

        }
        //drawingCanvas.destroy();
        return bitmap;
    }


    @SuppressLint("WrongThread")
    public void generateBitmapFile() {
        SyncTileBitmap bitmap = generateBitmap();


        try {
            OutputStream outStream = imageFile.openW();
            bitmap.getAndroidBitmap().compress(Bitmap.CompressFormat.PNG, 90, outStream);
            outStream.close();
            OldAppBroadcaster.broadcast(getContext(),
                    AppBroadcaster.FILE_CHANGED_ONDISK,
                    imageFile.toString(),
                    getClass().getName());

        } catch (Exception e) {
            AppLog.e(getContext(), e);
        }

        bitmap.free();
    }



    public boolean isReady() {
        return provider.isReadyAndLoaded();
    }


    @Override
    public void onDestroy() {
        provider.onDetached();
        super.onDestroy();
    }

}

File: aat-android/src/main/kotlin/ch/bailu/aat/map/mapsforge/MapsForgePreview.kt
package ch.bailu.aat.map.mapsforge

import android.content.Context
import ch.bailu.aat.map.MapDensity
import ch.bailu.aat_lib.app.AppContext
import ch.bailu.aat_lib.app.AppGraphicFactory
import ch.bailu.aat_lib.broadcaster.AppBroadcaster
import ch.bailu.aat_lib.gpx.information.GpxInformation
import ch.bailu.aat_lib.gpx.information.InfoID
import ch.bailu.aat_lib.lib.color.ColorInterface
import ch.bailu.aat_lib.logger.AppLog
import ch.bailu.aat_lib.map.layer.gpx.GpxDynLayer
import ch.bailu.aat_lib.map.tile.MapTileInterface
import ch.bailu.aat_lib.map.tile.MapsForgeTileLayer
import ch.bailu.aat_lib.map.tile.TileProvider
import ch.bailu.aat_lib.map.tile.source.CacheOnlySource
import ch.bailu.aat_lib.map.tile.source.MapsForgeSource
import ch.bailu.aat_lib.map.tile.source.Source
import ch.bailu.aat_lib.preferences.map.SolidMapTileStack
import ch.bailu.aat_lib.preferences.map.SolidRenderTheme
import ch.bailu.aat_lib.service.cache.DownloadSource
import ch.bailu.aat_lib.service.directory.MapPreviewInterface
import ch.bailu.foc.Foc
import org.mapsforge.core.graphics.Bitmap
import org.mapsforge.core.model.BoundingBox
import org.mapsforge.core.model.Dimension
import org.mapsforge.core.model.MapPosition
import org.mapsforge.core.model.Point
import org.mapsforge.core.model.Rotation
import org.mapsforge.map.util.LayerUtil
import org.mapsforge.map.util.MapPositionUtil
import org.mapsforge.map.view.FrameBuffer
import org.mapsforge.map.view.FrameBufferHA3

class MapsForgePreview(context: Context, private val appContext: AppContext, info: GpxInformation, out: Foc):
    MapsForgeViewBase(appContext, context, MapsForgePreview::class.java.simpleName, MapDensity()),
    MapPreviewInterface {

    private val imageFile: Foc
    private val provider: TileProvider
    private val mapPosition: MapPosition
    private val mapBounding: BoundingBox
    private val tlPoint: Point

    init {
        layout(0, 0, BITMAP_SIZE, BITMAP_SIZE)
        model.mapViewDimension.dimension = DIM
        imageFile = out
        provider = TileProvider(appContext, getSource(SolidRenderTheme(appContext.mapDirectories.createSolidDirectory(), appContext)))

        val tileLayer = MapsForgeTileLayer(appContext.services, provider, appContext.tilePainter)
        add(tileLayer)

        val gpxLayer = GpxDynLayer(appContext.storage, getMContext(), appContext.services)
        add(gpxLayer)
        attachLayers()
        gpxLayer.onContentUpdated(InfoID.FILE_VIEW, info)
        frameBounding(info.getGpxList().getDelta().getBoundingBox())
        mapPosition = model.mapViewPosition.mapPosition

        val tileSize = model.displayModel.tileSize
        mapBounding = boundingBox
        tlPoint = MapPositionUtil.getTopLeftPoint(mapPosition, DIM, tileSize)
        preLoadTiles()
    }

    private fun preLoadTiles() {
        val tilePositions = LayerUtil.getTilePositions(
            mapBounding,
            mapPosition.zoomLevel, tlPoint,
            model.displayModel.tileSize
        )
        provider.onAttached()
        provider.preload(tilePositions)
    }

    /**
     *
     * Begin of "prevent MapView from drawing" hack
     * FIXME:
     * This hack prevents the map view from calling the layers draw() function.
     * The correct implementation would be to port the preview generator away from the MapView and
     * just use the MapViews model.
     */
    override fun getFrameBuffer(): FrameBuffer {
        return object : FrameBufferHA3(
            model.frameBufferModel,
            model.displayModel,
            AppGraphicFactory.instance()
        ) {
            override fun getDrawingBitmap(): Bitmap? {
                return null
            }
        }
    }

    override fun repaint() {}
    override fun requestRedraw() {}

    /**
     * End of "prevent MapView from drawing" hack
     */

    private fun generateBitmap(): MapTileInterface {
        val bitmap = appContext.createMapTile()

        bitmap.set(BITMAP_SIZE, false)
        if (bitmap.isLoaded()) {
            val canvas = bitmap.getCanvas()
            bitmap.getBitmap()?.setBackgroundColor(ColorInterface.BLACK)
            for (layer in layerManager.layers) {
                layer.draw(mapBounding, mapPosition.zoomLevel, canvas, tlPoint, Rotation.NULL_ROTATION)
            }
        }
        return bitmap
    }

    override fun generateBitmapFile() {
        val bitmap = generateBitmap()
        try {
            imageFile.openW()?.use { out ->
                bitmap.getBitmap()?.compress(out)
                appContext.broadcaster.broadcast(
                    AppBroadcaster.FILE_CHANGED_ONDISK,
                    imageFile.path,
                    javaClass.name
                )
            }
        } catch (e: Exception) {
            AppLog.e(this, e)
        }
        bitmap.free()
    }

    override fun isReady(): Boolean {
        return provider.isReadyAndLoaded
    }

    override fun onDestroy() {
        provider.onDetached()
        super.onDestroy()
    }

    companion object {
        private const val BITMAP_SIZE = 128
        private val DIM = Dimension(BITMAP_SIZE, BITMAP_SIZE)
        private val MAPNIK: Source = CacheOnlySource(DownloadSource.MAPNIK)

        private fun getSource(stheme: SolidRenderTheme): Source {
            val tiles = SolidMapTileStack(stheme)
            val enabled = tiles.getEnabledArray()
            if (enabled[0]) {
                val theme = stheme.getValueAsString()
                val mfs = MapsForgeSource(theme)
                return CacheOnlySource(mfs)
            }
            return MAPNIK
        }
    }
}
